### PR TITLE
[SLP][REVEC] Fix insertelement has multiple uses.

### PR DESCRIPTION
**Pull request by HanKuanChen**

### CODE REVIEW

**Comment by RSilicon** [comment]
Is this fixing a miscompilation or is this an enhancement?

**Comment by HanKuanChen** [comment]
> Is this fixing a miscompilation or is this an enhancement?

This is a bug fix **only** when REVEC is enabled.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -6402,6 +6402,14 @@ BoUpSLP::TreeEntry::EntryState BoUpSLP::getScalarsVectorizationState(
       return TreeEntry::NeedToGather;
     }
 
+    if (any_of(VL, [&SourceVectors](Value *V) {
+          if (SourceVectors.contains(V))
+            return !V->hasOneUse();
+          // The last InsertElement can have multiple uses.
+          return false;
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:6409]: `return SourceVectors.contains(V) && !V->hasOneUse();`

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -6402,6 +6402,14 @@ BoUpSLP::TreeEntry::EntryState BoUpSLP::getScalarsVectorizationState(
...
+        }))
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:6410]: Add LLVM_DEBUG message.

Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp]: Need to add a check that this only REVEC specific

**Review by alexey-bataev — APPROVED** [review]
LG

**Pull request merged**